### PR TITLE
InvalidLinkBear: Split some parts into URLBear

### PR DESCRIPTION
**Pull request by refeed**
- Split some parts of InvalidLinkBear to URLBear.
Some methods and functions that related to URL finding
and status code checking are moved to URLBear.

- Assign URLBear to InvalidLinkBear's BEAR_DEPS,
so InvalidLinkBear just control the results of URLBear.

- Replace link_context from enum to use LINK_CONTEXT flags.
`link_context` that uses coalib.misc.Enum.enum() isn't pickleable.
So I replaced it by using `aenum`'s Flag.

- Change some InvalidLinkBear testcases to comply with URLBear.

- Change MementoBear to use URLBear instead of InvalidLinkBear.

- Delete MementoBear `link_ignore_list` setting, since it has been
defined in URLBear.

- Create URLBearTest.py.

Closes https://github.com/coala/coala-bears/issues/1871

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: tests/general/InvalidLinkBearTest.py [hunk 1]
```diff
@@ -404,7 +411,7 @@ def test_links_to_ignore(self):
                            'http://www.notexample.com/404',
                            'http://exampe.com/404',
                            'http://example.co.in/404'
-                          ]
+                           ]
```
Comment (jayvdb) [inline tests/general/InvalidLinkBearTest.py]: unrelated change?  should be a separate commit, or a 'newcomer' issue if it isnt critical.

File: tests/general/MementoBearTest.py [hunk 2]
```diff
@@ -142,13 +143,13 @@ class MementoBearTest(LocalBearTestHelper):
     """
 
     def setUp(self):
-        self.mb_check_prerequisites = MementoBear.check_prerequisites
+        self.mb_check_prerequisites = URLBear.check_prerequisites
```
Comment (jayvdb) [inline tests/general/MementoBearTest.py]: `mb_` prefix is no longer appropriate

File: tests/general/MementoBearTest.py [hunk 3]
```diff
@@ -283,9 +284,6 @@ def test_settings_follow_redirects(self):
 
     def test_links_to_ignore(self):
         valid_file = """
-        # web.archive.org should be ignored
-        http://web.archive.org/web/*/spiegel.com
```
Comment (jayvdb) [inline tests/general/MementoBearTest.py]: Why removed?

File: bears/general/MementoBear.py [hunk 4]
```diff
@@ -86,55 +62,40 @@ def run(self, filename, file,
...
-                             'archived.'
-                             % (url, url)),
-                            file=filename,
-                            line=line_number,
-                            severity=RESULT_SEVERITY.INFO
-                        )
+        for result in dependency_results.get(URLBear.name, []):
+            line_number, link, code, context = result.contents
+
+            if code and 200 <= code < 400:
```
Comment (jayvdb) [inline bears/general/MementoBear.py]: You've changed the indent levels of the code, resulting in lots of lines changing, which makes review harder.
This can be avoided here, by using

`if not (code and 200 <= code < 400): continue`

Comment (refeed) [inline bears/general/MementoBear.py]: thanks

File: tests/general/MementoBearTest.py (hunk 3 above)
Comment (refeed) [inline tests/general/MementoBearTest.py]: because I've remove `DEFAULT_IGNORE` in `MementoBear` https://github.com/coala/coala-bears/pull/1889/files#diff-fe065107cab5427f5ff5a6a80e74788fL25

Comment (jayvdb) [inline tests/general/MementoBearTest.py]: Ok, when this is merged, create a new issue to create a new default ignore list which is only used by MementoBear

**Comment by jayvdb** [comment]
`git blame -CC bears/general/URLBear.py` is looking quite good, with a lot of the lines retaining their old authorship :+1:

File: bears/general/URLBear.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    LANGUAGES = {'All'}
+    REQUIREMENTS = {PipRequirement('requests', '2.12'),
+                    PipRequirement('aenum', '2.0.8')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Documentation'}
+
+    # IP Address of www.google.com
+    check_connection_url = 'http://216.58.218.174'
```
Comment (jayvdb) [inline bears/general/URLBear.py]: when this is merged, create an issue to make `URLBear.check_connection_url` a setting instead of a hidden constant

Comment (refeed) [inline bears/general/URLBear.py]: okay :+1:

File: bears/general/URLBear.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+from coalib.results.HiddenResult import HiddenResult
+from coalib.settings.Setting import typed_list, typed_dict
+from coalib.parsing.Globbing import fnmatch
+
+
+class LINK_CONTEXT(Flag):
+    no_flags = 0
+    xml_namespace = 1
+    pip_vcs_url = 2
+    all_flags = 3
```
Comment (jayvdb) [inline bears/general/URLBear.py]: https://bitbucket.org/stoneleaf/aenum says you can use bitwise comparison, which means this `all_flags` shouldnt be necessary.

Comment (jayvdb) [inline bears/general/URLBear.py]: Also, they are not able to combined. A pip context is *not* possible in a xml namespace context.

Comment (refeed) [inline bears/general/URLBear.py]: > Also, they are not able to combined. A pip context is not possible in a xml namespace context.

Right!

File: tests/general/MementoBearTest.py (hunk 3 above)
Comment (refeed) [inline tests/general/MementoBearTest.py]: ok

**Comment by jayvdb** [comment]
ack a71b222

**Comment by jayvdb** [comment]
reack 012b7d1

**Comment by jayvdb** [comment]
@rultor merge

File: bears/general/URLBear.py (hunk 5 above)
Comment (refeed) [inline bears/general/URLBear.py]: https://github.com/coala/coala-bears/issues/1894

File: tests/general/MementoBearTest.py (hunk 3 above)
Comment (refeed) [inline tests/general/MementoBearTest.py]: https://github.com/coala/coala-bears/issues/1895
